Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp

```diff
@@ -0,0 +1,544 @@
+/*
+ * Copyright (c) 2023, Lucas Chollet <lucas.chollet@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "TIFFLoader.h"
+#include <AK/Debug.h>
+#include <AK/Endian.h>
+#include <AK/String.h>
+
+namespace Gfx {
+
+class TIFFLoadingContext {
+public:
+    enum class State {
+        NotDecoded = 0,
+        Error,
+        HeaderDecoded,
+        FrameDecoded,
+    };
+
+    template<OneOf<u32, i32> x32>
+    struct Rational {
+        using Type = x32;
+        x32 numerator;
+        x32 denominator;
+    };
+
+    TIFFLoadingContext(NonnullOwnPtr<FixedMemoryStream> stream)
+        : m_stream(move(stream))
+    {
+    }
+
+    ErrorOr<void> decode_image_header()
+    {
+        TRY(read_image_file_header());
+        TRY(read_next_image_file_directory());
+
+        m_state = State::HeaderDecoded;
+        return {};
+    }
+
+    ErrorOr<void> decode_frame()
+    {
+        auto maybe_error = decode_frame_impl();
+
+        if (maybe_error.is_error()) {
+            m_state = State::Error;
+            return maybe_error.release_error();
+        }
+
+        return {};
+    }
+
+    IntSize size() const
+    {
+        return m_size;
+    }
+
+    State state() const
+    {
+        return m_state;
+    }
+
+    RefPtr<Bitmap> bitmap() const
+    {
+        return m_bitmap;
+    }
+
+private:
+    enum class ByteOrder {
+        LittleEndian,
+        BigEndian,
+    };
+
+    enum class Type {
+        Byte = 1,
+        ASCII = 2,
+        Short = 3,
```

Comment:
Let's call this UnsignedShort. The spec calls it "Short", but it also says it's unsigned so let's be explicit about it (also for Long, Rational).

(I'd also say `SignedLong` instead of `SLong` etc below.)